examples/pytorch/rgcn/utils.py

```diff
@@ -334,3 +343,326 @@ def calc_mrr(embedding, w, train_triplets, valid_triplets, test_triplets, hits=[
+    raw_reverse_sugb = {}
+    raw_reverse_subg_etype = {}
+    print(num_rels)
+
+    settype = 0
+    for edge_list in edge_lists:
+        s_l, d_l, st_l, r_l, dt_l = edge_list
+        for i in range(s_l.shape[0]):
+            s = s_l[i]
+            d = d_l[i]
+            st = st_l[i]
+            r = r_l[i]
+            dt = dt_l[i]
+            assert r < num_rels
+            s_type = str(st)
+            d_type = str(dt)
+            r_type = str(r)
+            e_type = (s_type, r_type, d_type)
+
+            if raw_subg.get(e_type, None) is None:
+                raw_subg[e_type] = ([], [])
+                raw_subg_eset[e_type] = []
+                raw_subg_etype[e_type] = []
+            raw_subg[e_type][0].append(s)
+            raw_subg[e_type][1].append(d)
+            raw_subg_eset[e_type].append(settype)
+            raw_subg_etype[e_type].append(r)
+
+            if reverse is True:
+                r_type = str(r + num_rels)
+                re_type = (d_type, r_type, s_type)
+                if raw_reverse_sugb.get(re_type, None) is None:
+                    raw_reverse_sugb[re_type] = ([], [])
+                    raw_reverse_subg_etype[re_type] = []
+                raw_reverse_sugb[re_type][0].append(d)
+                raw_reverse_sugb[re_type][1].append(s)
+                raw_reverse_subg_etype[re_type].append(r + num_rels)
+        settype += 1
+
+    subg = []
+    fg_s = []
+    fg_d = []
+    fg_etype = []
+    fg_settype = []
+    for e_type, val in raw_subg.items():
+        s, d = val
+        s = np.asarray(s)
+        d = np.asarray(d)
+        etype = raw_subg_etype[e_type]
+        etype = torch.tensor(etype).long()
+        settype = raw_subg_eset[e_type]
+        settype = torch.tensor(settype).long()
+
+        fg_s.append(s)
+        fg_d.append(d)
+        fg_etype.append(etype)
+        fg_settype.append(settype)
+
+    for e_type, val in raw_reverse_sugb.items():
+        s, d = val
+        s = np.asarray(s)
+        d = np.asarray(d)
+        etype = raw_reverse_subg_etype[e_type]
+        etype = torch.tensor(etype).long()
+        settype = torch.full((s.shape[0],), -1).long()
+
+        fg_s.append(s)
+        fg_d.append(d)
+        fg_etype.append(etype)
+        fg_settype.append(settype)
+
+    s = np.concatenate(fg_s)
+    d = np.concatenate(fg_d)
+    g = dgl.graph((s, d), num_nodes=num_nodes)
+    g.edata['etype'] = torch.cat(fg_etype)
+    g.edata['set'] = torch.cat(fg_settype)
+    g.ndata['ntype'] = torch.from_numpy(node_types).long()
+
+    return g
+
+#######################################################################
+#
+# Build Heterograph in Homogeneous Graph for Link Prediction
+#
+#######################################################################
+def build_heterograph_in_homogeneous_from_triplets(num_nodes, num_rels, edge_lists, reverse=True):
+    """ Create a DGL homogeneous graph with heterograph info stored as node or edge features.
+    """
```

Comment:
What's the difference between this function and the one above?

Reply:
One for node with type and another node has no type.